Add Parallelization to SolutionInvalidity (#23198)
idaholab/moose: framework/doc/content/source/problems/SolutionInvalidity.md

# SolutionInvalidity

## Description

> The [/SolutionInvalidity.md] object holds solution invalid warning information for MOOSE. With this object, you can mark a solution as "invalid" and output which object and how many times the warning occurs. An invalid solution means that the solution somehow does not satisfy requirements such as a value being out of the bounds of a correlation.  Solutions are allowed to be invalid *during* the nonlinear solve - but are not allowed to be invalid once it converges. A "converged" solution that is marked as invalid will cause MOOSE to behave as if the solution did NOT converge - including cutting back timesteps, etc.

Reviewer: ```suggestion
The [/SolutionInvalidity.md] object holds solution invalid warning information for MOOSE. With this object, you can mark a solution as "invalid" and output which object and how many times the warning occurs. An invalid solution means that the solution somehow does not satisfy requirements, such as a value being out of the bounds of a correlation.  Solutions are allowed to be invalid *during* the nonlinear solve, but are not allowed to be invalid once it converges. A "converged" solution that is marked as invalid will cause MOOSE to behave as if the solution did NOT converge - including cutting back timesteps, etc.
```

> To declare a solution as "invalid", use the macro in the following code to mark it. The user could provide a short description of the object(optional) and a message to describe the invalidity(+required+):

Reviewer: ```suggestion
To declare a solution as "invalid", use the macro in the following code to mark it. The user can provide a short description of the object (optional) and a message to describe the invalidity (+required+):
```

!listing /test/src/materials/NonsafeMaterial.C  re=\s+if \(_input_diffusivity > _threshold\).*?\}

!alert tip

> It is recommended to have a short description for each object especially when you want to mark multiple types of invalid solutions within one object.

Reviewer: ```suggestion
It is recommended to have a short description for each object, especially when you want to mark multiple types of invalid solutions within one object.
```

> If any solution invalidity is detected during the solve, a summary table of solution invalid warnings will be generated and reported at the end of each time step as below:

Reviewer: ```suggestion
If any solution invalidity is detected during the solve, a summary table of solution invalid warnings will be generated and reported at the end of each time step. An example of this output is shown below:
```

```
Solution Invalid Warnings:
------------------------------------------------------------------------------------------------------------------------------

> |                 Object                 | Current | Timestep | Total |                        Message                       |

Reviewer 2: what is "Current" as opposed to Timestep ?

------------------------------------------------------------------------------------------------------------------------------
| NonsafeMaterial::Diffusivity check     |      16 |       64 |    64 | The diffusivity is greater than the threshold value! |
| NonsafeMaterial::Second thing to check |      16 |       64 |    64 | Extra invalid thing!                                 |
------------------------------------------------------------------------------------------------------------------------------
```

1. *Object*: shows the moose object name registered for the solution invalid detection and an optional discription
2. *Current*: shows the number of solution invalid warnings for the latest iteration
3. *Timestep*: shows the number of solution invalid warnings for one time step
4. *Total*: shows the total number of soluton invalid warnings for the solve
5. *Message*: shows the decription of the solution invalidity

> This Solution Invalid Warning table can be silenced by setting [!param](/Problem/FEProblem/allow_invalid_solution) to 'true'. Then the converged solution will be allowed even there are still solution invalid warnings, but a message will be generated in the end of the calculation as a reminder:

Reviewer: ```suggestion
This Solution Invalid Warnings table can be silenced by setting [!param](/Problem/FEProblem/allow_invalid_solution) to 'true'. Then the converged solution will be allowed even if there are still solution invalid warnings, but a message will be generated in the end of the calculation as a reminder:
```

```
*** Warning ***
The Solution Invalidity warnings are detected but silenced! Use Problem/allow_invalid_solution=false to activate
```

The Solution Invalid Warning can also be printed out immediately after it is detected by setting [!param](/Problem/FEProblem/immediately_print_invalid_solution) to `true`.

> 

Reviewer 2: Add a note about how the number of warnings can map to the number of quadrature points / nodes /elements affected or could be global. The message should provide enough information to the user to take remedial action

Author: Comment has been addressed

# SolutionInvalidity

## Description

> The [/SolutionInvalidity.md] object holds solution invalid warning information for MOOSE. With this object, you can mark a solution as "invalid" and output which object and how many times the warning occurs. An invalid solution means that the solution somehow does not satisfy requirements, such as a value being out of the bounds of a correlation.  Solutions are allowed to be invalid *during* the nonlinear solve, but they are not allowed to be invalid once it converges. A "converged" solution that is marked as invalid will cause MOOSE to behave as if the solution did NOT converge - including cutting back timesteps, etc.

Reviewer 2: ```suggestion
The [/SolutionInvalidity.md] object holds solution invalid warning information for MOOSE. This object allows you to mark a solution as "invalid" under certain conditions and output in which objects and how many times the conditions occurs. An invalid solution means that the solution somehow does not satisfy requirements, such as a value being out of the bounds of a correlation.  Solutions are allowed to be invalid *during* the nonlinear solve, but they are not allowed to be invalid once the solve has converged. A "converged" solution that is marked as invalid will cause MOOSE to behave as if the solution did NOT converge - including cutting back timesteps, etc.
```

> To declare a solution as "invalid", use the macro in the following code to mark it. The user can provide a message to describe the invalidity:

Reviewer 2: ```suggestion
To declare a solution as "invalid", use the macro in the following code to mark it. The user can provide a message to describe the reason for invalidity:
```

!listing /test/src/materials/NonsafeMaterial.C  re=\s+if \(_input_diffusivity > _threshold\).*?\}

!alert tip
It is recommended to have a unique message for each invalidity, especially when you want to mark multiple types of invalid solutions within one object.

If any solution invalidity is detected during the solve, a summary table of solution invalid warnings will be generated and reported at the end of each time step. An example of this output is shown below:

```
Solution Invalid Warnings:
------------------------------------------------------------------------------------------------------

> |     Object      | Latest | Timestep | Total |                        Message                       |

Reviewer 2: I still dont like Latest. 
This is the latest residual computation right? so can we say "Converged" instead?

Author: The comment has been addressed

------------------------------------------------------------------------------------------------------
| NonsafeMaterial |     16 |       64 |    64 | The diffusivity is greater than the threshold value! |
| NonsafeMaterial |     16 |       64 |    64 | Extra invalid thing!                                 |

> ------------------------------------------------------------------------------------------------------

Reviewer 2: if we have 0 warnings during the timestep, does the warning get silenced?

Author: Yes, there will not be a summary table in the end.
![Screenshot 2023-02-12 at 8 42 05 PM](https://user-images.githubusercontent.com/118846840/218365449-880a61da-1155-4401-b116-8c66c43670bb.png)

Reviewer 2: What if there s a warning in timestep 1 and no warning in timestep 2?

Author: The summary table is only printed out when solution invalid is found as coded in NonlinearSystem.C
![Screenshot 2023-02-12 at 9 37 44 PM](https://user-images.githubusercontent.com/118846840/218371817-9cfe6451-1441-4f57-a399-3eeb2f81ce42.png)

```

1. *Object*: shows the moose object name registered for the solution invalid detection and an optional description

> 2. *Latest*: shows the number of solution invalid warnings for the latest iteration

Reviewer 2: ```suggestion
2. *Latest*: shows the number of solution invalid warnings for the latest iteration, at convergence
```

> 3. *Timestep*: shows the number of solution invalid warnings for the latest time step

Reviewer 2: ```suggestion
3. *Timestep*: shows the number of solution invalid warnings for the latest time step over all iterations
```

4. *Total*: shows the total number of solution invalid warnings for the entire simulation
5. *Message*: shows the description of the solution invalidity

> SolutionInvalidity will check the value of the quadrature points during each linear and nonlinear iterations. When the solution is converged, the number of solution invalid warnings is reported for the latest iteration(linear/nonlinear), the last time step and the entire simulation.

Reviewer 2: ```suggestion
When the solution is converged, the number of solution invalid warnings is reported for the latest iteration (linear/nonlinear), the last time step (over all solver iterations) and the entire simulation.
SolutionInvalidity warnings can come from numerous objects, some executed once and some executed on every quadrature point. The number of occurrences reported reflects this.
```

| NonsafeMaterial |        16 |       64 |    64 | The diffusivity is greater than the threshold value! |
| NonsafeMaterial |        16 |       64 |    64 | Extra invalid thing!                                 |
---------------------------------------------------------------------------------------------------------
```

1. *Object*: shows the moose object name registered for the solution invalid detection and an optional description
2. *Converged*: shows the number of solution invalid warnings for the latest iteration, at convergence
3. *Timestep*: shows the number of solution invalid warnings for the latest time step over all iterations
4. *Total*: shows the total number of solution invalid warnings for the entire simulation
5. *Message*: shows the description of the solution invalidity

When the solution is converged, the number of solution invalid warnings is reported for the latest iteration (linear/nonlinear), the last time step (over all solver iterations) and the entire simulation.
SolutionInvalidity warnings can come from numerous objects, some executed once and some executed on every quadrature point. The number of occurrences reported reflects this.

> This Solution Invalid Warnings table can be silenced by setting [!param](/Problem/FEProblem/allow_invalid_solution) to 'true'. Then the converged solution will be accepted even if there are still solution invalid warnings, but a message will be generated in the end of the calculation as a reminder:

Reviewer 2: ```suggestion
This Solution Invalid Warnings table can be silenced by setting [!param](/Problem/FEProblem/allow_invalid_solution) to 'true'. Then the converged solution will be accepted even if there are still solution invalid warnings, but a message will be generated at the end of the calculation as a reminder:
```
